tests/restore_test.go

```diff
@@ -0,0 +1,683 @@
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			deleteVM(vm)
+		})
+
+		Context("and no snapshot", func() {
+			It("should reject restore", func() {
+				restore := createRestoreDef(vm, "foobar")
+
+				_, err := virtClient.VirtualMachineRestore(vm.Namespace).Create(restore)
+				Expect(err).To(HaveOccurred())
+				Expect(err.Error()).To(ContainSubstring("VirtualMachineSnapshot \"foobar\" does not exist"))
+			})
+		})
+
+		Context("and good snapshot exists", func() {
+			var err error
+			var snapshot *snapshotv1.VirtualMachineSnapshot
+			var webhook *admissionregistrationv1beta1.ValidatingWebhookConfiguration
+
+			BeforeEach(func() {
+				snapshot = createSnapshot(vm)
+			})
+
+			AfterEach(func() {
+				deleteSnapshot(snapshot)
+				if webhook != nil {
+					deleteWebhook(webhook)
+				}
+			})
+
+			It("should successfully restore", func() {
+				var origSpec *v1.VirtualMachineSpec
+
+				Eventually(func() bool {
+					var updatedVM *v1.VirtualMachine
+					vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					origSpec = vm.Spec.DeepCopy()
+					Expect(origSpec.Template.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory]).To(Equal(resource.MustParse("64M")))
+
+					vm.Spec.Template.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("128M")
+					updatedVM, err = virtClient.VirtualMachine(vm.Namespace).Update(vm)
+					if errors.IsConflict(err) {
+						return false
+					}
+					vm = updatedVM
+					Expect(err).ToNot(HaveOccurred())
+					Expect(vm.Spec.Template.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory]).To(Equal(resource.MustParse("128M")))
+					return true
+				}, 180*time.Second, time.Second).Should(BeTrue())
+
+				restore := createRestoreDef(vm, snapshot.Name)
+
+				restore, err = virtClient.VirtualMachineRestore(vm.Namespace).Create(restore)
+				Expect(err).ToNot(HaveOccurred())
+
+				restore = waitRestoreComplete(restore, vm)
+				Expect(restore.Status.Restores).To(HaveLen(0))
+				Expect(restore.Status.DeletedDataVolumes).To(HaveLen(0))
+
+				vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &metav1.GetOptions{})
+				Expect(vm.Spec).To(Equal(*origSpec))
+
+				deleteRestore(restore)
+			})
+
+			It("should reject restore if VM running", func() {
+				patch := []byte("[{ \"op\": \"replace\", \"path\": \"/spec/running\", \"value\": true }]")
+				vm, err := virtClient.VirtualMachine(vm.Namespace).Patch(vm.Name, types.JSONPatchType, patch)
+				Expect(err).ToNot(HaveOccurred())
+
+				restore := createRestoreDef(vm, snapshot.Name)
+
+				restore, err = virtClient.VirtualMachineRestore(vm.Namespace).Create(restore)
+				Expect(err).To(HaveOccurred())
+				Expect(err.Error()).To(ContainSubstring(fmt.Sprintf("VirtualMachine %q is running", vm.Name)))
+			})
+
+			It("should reject restore if another in progress", func() {
```

Comment:
That's a pretty interesting idea to use a webhook to gate progress on something for a test.